libs/wingc/src/type_check.rs

```diff
@@ -1086,6 +1089,30 @@ impl Types {
 
 		self.resource_base_type.unwrap()
 	}
+
+	/// Stores the type of a given expression node.
+	pub fn assign_type_to_expr(&mut self, expr: &Expr, type_: TypeRef) {
+		let expr_idx = expr.id;
+		if self.type_for_expr.len() <= expr_idx {
+			self.type_for_expr.resize_with(expr_idx + 1, || None);
+		}
+		self.type_for_expr[expr_idx] = Some(type_);
+	}
+
+	/// Obtain the type of a given expression node. Returns None if the expression has not been type checked yet. If
+	/// this is called after type checking, it should always return Some.
+	pub fn get_expr_type(&self, expr: &Expr) -> Option<TypeRef> {
+		let expr_idx = expr.id;
+		if self.type_for_expr.len() <= expr_idx {
+			return None;
+		}
+		self.type_for_expr[expr_idx]
```

Comment:
nit: this might work and be cleaner
```suggestion
		self.type_for_expr.get(expr_idx).map(|t| *t);
```

Reply:
nice catch :) That snippet produced an Option<Option<TypeRef>>, but using `.map().flatten()` aka `.and_then()` fixed it